pkg/asset/machines/aws/master.go

```diff
@@ -2,72 +2,106 @@
+	platform := config.Platform.AWS
+	mpool := pool.Platform.AWS
+	azs := mpool.Zones
 
-// MasterInstance contains information specific to each
-// master machine instance to create.
-type MasterInstance struct {
-	AvailabilityZone string
-}
+	total := int64(1)
+	if pool.Replicas != nil {
+		total = *pool.Replicas
+	}
+	numOfAZs := int64(len(azs))
+	for idx, az := range azs {
+		replicas := int32(total / numOfAZs)
+		if int64(idx) < total%numOfAZs {
+			replicas++
+		}
+
+		tags := map[string]string{
+			"tectonicClusterID":                                  config.ClusterID,
+			fmt.Sprintf("kubernetes.io/cluster/%s", clustername): "owned",
+		}
+		for k, v := range platform.UserTags {
+			tags[k] = v
+		}
+		ptags := []awsprovider.TagSpecification{}
+		for k, v := range tags {
+			ptags = append(ptags, awsprovider.TagSpecification{Name: k, Value: v})
+		}
 
-// MasterMachineTmpl is a template for a list of master machines.
-var MasterMachineTmpl = template.Must(template.New("aws-master-machine").Parse(`
-{{- $c := . -}}
-kind: List
-apiVersion: v1
-metadata:
-  resourceVersion: ""
-  selfLink: ""
-items:
-{{- range $index,$instance := $c.Instances}}
-- apiVersion: cluster.k8s.io/v1alpha1
-  kind: Machine
-  metadata:
-    name: {{$c.ClusterName}}-master-{{$index}}
-    namespace: openshift-cluster-api
-    labels:
-      sigs.k8s.io/cluster-api-cluster: {{$c.ClusterName}}
-      sigs.k8s.io/cluster-api-machine-role: master
-      sigs.k8s.io/cluster-api-machine-type: master
-  spec:
-    providerConfig:
-      value:
-        apiVersion: aws.cluster.k8s.io/v1alpha1
-        kind: AWSMachineProviderConfig
-        ami:
-          id: {{$c.AMIID}}
-        instanceType: {{$c.Machine.InstanceType}}
-        placement:
-          region: {{$c.Region}}
-          availabilityZone: {{$instance.AvailabilityZone}}
-        subnet:
-          filters:
-          - name: "tag:Name"
-            values:
-            - "{{$c.ClusterName}}-master-{{$instance.AvailabilityZone}}"
-        iamInstanceProfile:
-          id: "{{$c.ClusterName}}-master-profile"
-        tags:
-{{- range $key,$value := $c.Tags}}
-          - name: "{{$key}}"
-            value: "{{$value}}"
-{{- end}}
-        securityGroups:
-          - filters:
-            - name: "tag:Name"
-              values:
-              - "{{$c.ClusterName}}_master_sg"
-        userDataSecret:
-          name: "master-user-data"
-    versions:
-      kubelet: ""
-      controlPlane: ""
-{{- end}}
-`))
+		provider := awsprovider.AWSMachineProviderConfig{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "aws.cluster.k8s.io/v1alpha1",
+				Kind:       "AWSMachineProviderConfig",
+			},
+			InstanceType:       mpool.InstanceType,
+			AMI:                awsprovider.AWSResourceReference{ID: &mpool.AMIID},
+			Tags:               ptags,
+			IAMInstanceProfile: &awsprovider.AWSResourceReference{ID: func(s string) *string { return &s }(fmt.Sprintf("%s-master-profile", clustername))},
```

Comment:
I think this would be a bit easier to read as a `pointerSprintf` helper or some such:

```go
IAMInstanceProfile: &awsprovider.AWSResourceReference{ID: pointerSprintf("%s-master-profile", clustername)},
```

Thoughts?

Reply:
don't want `pointerSprintf` if its only used once.
there is completely separate lib `k8s.io/utils/pointer` for this a go doesn't let you do `&func`

Reply:
> don't want pointerSprintf if its only used once.

Twice:

```console
$ git grep 'func(s string) \*string' origin/pr/573  |cat
origin/pr/573:pkg/asset/machines/aws/master.go:			IAMInstanceProfile: &awsprovider.AWSResourceReference{ID: func(s string) *string { return &s }(fmt.Sprintf("%s-master-profile", clustername))},
origin/pr/573:pkg/asset/machines/aws/worker.go:			IAMInstanceProfile: &awsprovider.AWSResourceReference{ID: func(s string) *string { return &s }(fmt.Sprintf("%s-worker-profile", clustername))},
```

> there is completely separate lib `k8s.io/utils/pointer` for this a go doesn't let you do `&func`

I'm fine vendoring that.  It probably doesn't have much in the way of secondary dependencies, right?